kernel/queue.c

```diff
@@ -136,11 +136,14 @@ static inline void z_vrfy_k_queue_cancel_wait(struct k_queue *queue)
 #endif
 
 static int32_t queue_insert(struct k_queue *queue, void *prev, void *data,
-			  bool alloc)
+			    bool alloc, bool is_append)
 {
-	k_spinlock_key_t key = k_spin_lock(&queue->lock);
 	struct k_thread *first_pending_thread;
+	k_spinlock_key_t key = k_spin_lock(&queue->lock);
 
+	if (is_append) {
```

Comment:
Might be simpler to use a NULL prev to signal this condition instead?  Not a big deal, but there are a lot of boilerplate changes here to add the extra argument, and the function already had five arguments.

Alternatively: split the function up into a must-be-locked core and a wrapper to manage the locking, then add a new "queue_append()" wrapper that selects an appropriate prev?

Reply:
I had the same thought, about using NULL prev to signal the append condition, but that will not work as the k_queue_prepend also sets prev to NULL thus the prepend would become append.

I do not really see the benefit of splitting the function, since the tail peek together with the is_append flag makes it simple, easy to understand and maintain, and also efficient as there is no need for extra function calls... But that is just my opinion.